Add coverage checks to property tests for `TokenMap.intersection`

jonathanknowles opened the pull request:
# Issue Number

ADP-997

# Pre-merge checks

- [x] Perform soak test of test suite to ensure no flakiness has been introduced.
    (✅ 1000 test runs with `--match "/Token map properties/Arithmetic/prop_intersection"` passed with no failures.)

# Overview

PR #2725 added a `TokenMap.intersection` function and property tests.

This PR:
- [x] adds coverage checks to the `prop_intersection_*` series of properties.
- [x] where necessary, adjusts property inputs in order to attain the required level of coverage.

jonathanknowles committed:
Add `ManyFolded` combinator to `Test.QuickCheck.Extra`.

This combinator provides a way to generate larger arbitrary values
through simple monoidal concatenation of smaller arbitrary values.

jonathanknowles committed:
Add coverage checks to `prop_intersection_associativity`.

Before this change, coverage of the intersection is not empty" condition
was only 1%.

jonathanknowles committed:
Add coverage checks to `prop_intersection_commutativity`.

jonathanknowles committed:
Add coverage checks to `prop_intersection_empty`.

jonathanknowles committed:
Add coverage checks to `prop_intersection_equality`.

jonathanknowles committed:
Add coverage checks to `prop_intersection_identity`.

jonathanknowles committed:
Add coverage checks to `prop_intersection_subset`.

sevanspowell reviewed (approved):
Looks good to me!

sevanspowell commented on lib/test-utils/src/Test/QuickCheck/Extra.hs:
```diff
@@ -51,3 +57,22 @@ shrinkInterleaved (a, shrinkA) (b, shrinkB) = interleave
+-- generate larger values of 'A', based on simple monoidal concatenation of
+-- multiple smaller values.
+--
+newtype ManyFolded a = ManyFolded
```
This is required because the available Arbitrary instance for TokenMap is something like this:

```
instance Arbitrary TokenMap where
    arbitrary = genTokenMapSmallRange
```

right?

I think I prefer Hedgehog's use of functions to Quickcheck's typeclass instances, but that's not the question here.

rvl replied:
I found this confusing @jonathanknowles.

Adding in `Blind` didn't help (couldn't `Show` be defined with `show . mconcat`?

Why can't we just use the QuickCheck `size` parameter? See [Coding Standards - QuickCheck](https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-the-size-parameter).

jonathanknowles replied:
Thanks @rvl. I agree with you that using the `size` parameter would be a better solution here.

I've added a commit which does this here: 86f48565e9822e3cb1d76721236e621c152e72e5

rvl reviewed (commented):
Thanks @jonathanknowles .
Coverage checks - good.
ManyFolded - unsure.

jonathanknowles committed:
Add functions `{gen,shrink}TokenQuantitySized`.

This allows the generation and shrinking of token quantities from a range
that depends on the size parameter.

jonathanknowles committed:
Add functions `{gen,shrink}TokenNameSized`.

This allows the generation and shrinking of token names from a range
that depends on the size parameter.

jonathanknowles committed:
Add functions `{gen,shrink}TokenPolicyIdSized`.

This allows the generation and shrinking of token policy identifiers
from a range that depends on the size parameter.

jonathanknowles committed:
Add functions `{gen,shrink}AssetIdSized`.

This allows the generation and shrinking of asset identifiers from a range
that depends on the size parameter.

jonathanknowles committed:
Add functions `{gen,shrink}TokenMapSized`.

This allows the generation and shrinking of token maps where the number
of entries, the range of asset identifiers, and the range of token
quantities all depend on the size parameter.

jonathanknowles committed:
Use sized generators for `TokenMapSpec.prop_intersection_*`.

In this change, we use sized generators for token maps instead of
relying on the `ManyFolded` combinator.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2756#discussion_r667625710
